### PR TITLE
add message handler for edgehub

### PR DESCRIPTION
**Pull request by gy95**
Signed-off-by: gy95 <guoyao17@huawei.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
introduce message handler to process some specific messages.
When we want to process some particular messages from cloud, we could call `RegisterHandler` function to register message handlers, so when the compliant messages come, we could process it. And this method is a non-intrusive way to add more message processing logics, we don't neet to add more branch cases directly to the blow block `swich case` anymore, and this makes codes more readability and elegant.

https://github.com/kubeedge/kubeedge/blob/9abf3aa6ee4df0a0746acd8507e1c9b18af59eb6/edge/pkg/edgehub/process.go#L47-L59
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by kevin-wangzefeng** [comment]
This is a refectoring work, I would suggest include change on existing code.
Otherwise it's no-op.

**Comment by gy95** [comment]
This message handler is better for processing messages which are only from `cloudcore` and not from inner modules of `edgecore`.

using this handler for devicetwin is not suitable, because meesages that need to be processed by `devicetwin` module, are not only from cloudcore, but also from eventbus etc. If devicetwin also register handlers, will make codes more complex

**Comment by gy95** [comment]
/kindly @fisherxu @kevin-wangzefeng ,PTAL, is this modification ok or not? if ok, i'll squalish the commits.

File: edge/pkg/edgehub/process.go [hunk 1]
```diff
@@ -53,26 +68,34 @@ func (eh *EdgeHub) dispatch(message model.Message) error {
...
 
+func (eh *EdgeHub) dispatch(message model.Message) error {
+	// handler for msg.
+	isProcessed, err := msghandler.ProcessHandler(message, eh.chClient)
+	if isProcessed {
+		return err
+	}
+
+	klog.Errorf("msg not processed due to not-supported, group: %v", message.GetGroup())
+	return fmt.Errorf("msg not processed due to not-supported, group: %v", message.GetGroup())
```
Comment (fisherxu) [inline edge/pkg/edgehub/process.go:96]: Will always return error?

File: edge/pkg/edgehub/common/msghandler/handler.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	for _, handle := range Handlers {
+		if handle.Filter(&message) {
+			err := handle.Process(&message, client)
+			if err != nil {
+				return true, fmt.Errorf("failed to handle message, error: %+v", err)
+			}
+			return true, nil
+		}
+	}
+	return false, nil
```
Comment (fisherxu) [inline edge/pkg/edgehub/common/msghandler/handler.go:44]:
```suggestion
	return true, nil
```
?

File: edge/pkg/edgehub/common/msghandler/handler.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+// ProcessHandler return true if handler filtered
+func ProcessHandler(message model.Message, client clients.Adapter) (bool, error) {
+	lock.RLock()
+	defer lock.RUnlock()
+	for _, handle := range Handlers {
+		if handle.Filter(&message) {
+			err := handle.Process(&message, client)
+			if err != nil {
+				return true, fmt.Errorf("failed to handle message, error: %+v", err)
```
Comment (fisherxu) [inline edge/pkg/edgehub/common/msghandler/handler.go:39]:
```suggestion
				return false, fmt.Errorf("failed to handle message, error: %+v", err)
```
?

File: edge/pkg/edgehub/process.go [hunk 4]
```diff
@@ -53,26 +68,34 @@ func (eh *EdgeHub) dispatch(message model.Message) error {
...
+		beehiveContext.SendToGroup(md, *message)
 	}
 	return nil
 }
 
+func (eh *EdgeHub) dispatch(message model.Message) error {
+	// handler for msg.
+	isProcessed, err := msghandler.ProcessHandler(message, eh.chClient)
+	if isProcessed {
+		return err
```
Comment (fisherxu) [inline edge/pkg/edgehub/process.go:92]:
```suggestion
		return nil
```
?

File: edge/pkg/edgehub/common/msghandler/handler.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+// RegisterHandler register msg handler.
+func RegisterHandler(handler Handler) {
+	lock.Lock()
+	defer lock.Unlock()
+	Handlers = append(Handlers, handler)
+}
+
+// ProcessHandler return true if handler filtered
+func ProcessHandler(message model.Message, client clients.Adapter) (bool, error) {
```
Comment (fisherxu) [inline edge/pkg/edgehub/common/msghandler/handler.go:32]: Do we need the bool here?

File: edge/pkg/edgehub/common/msghandler/handler.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+func init() {
+	Handlers = make([]Handler, 0)
+}
+
+// RegisterHandler register msg handler.
+func RegisterHandler(handler Handler) {
+	lock.Lock()
+	defer lock.Unlock()
+	Handlers = append(Handlers, handler)
```
Comment (fisherxu) [inline edge/pkg/edgehub/common/msghandler/handler.go:28]: Will we have many handler in future? @wackxu

File: edge/pkg/edgehub/process.go [hunk 7]
```diff
@@ -37,11 +38,25 @@ func (eh *EdgeHub) initial() (err error) {
...
+	msghandler.RegisterHandler(handler)
+}
+
+type defaultHandler struct {
+}
+
+func (*defaultHandler) Filter(message *model.Message) bool {
+	group := message.GetGroup()
+	return group == messagepkg.ResourceGroupName || group == messagepkg.TwinGroupName ||
+		group == messagepkg.FuncGroupName || group == messagepkg.UserGroupName
```
Comment (fisherxu) [inline edge/pkg/edgehub/process.go:56]: Is there some message that filter failed? @wackxu

File: edge/pkg/edgehub/process.go (hunk 4 above)
Comment (gy95) [reply edge/pkg/edgehub/process.go:92]: if been processed, we need to return nil if no-error, return error if error happens, so just return `err`, and it contains nil or error circumanstances.

File: edge/pkg/edgehub/common/msghandler/handler.go (hunk 3 above)
Comment (gy95) [reply edge/pkg/edgehub/common/msghandler/handler.go:39]: this bool return value represents for whether message been processed by processed processors

File: edge/pkg/edgehub/common/msghandler/handler.go (hunk 2 above)
Comment (gy95) [reply edge/pkg/edgehub/common/msghandler/handler.go:44]: no, it means not matched processors and don't be processed.

File: edge/pkg/edgehub/process.go (hunk 1 above)
Comment (gy95) [reply edge/pkg/edgehub/process.go:96]: all supported messages will be processed by processors. So if the codes run here, it means that it's not supported, so just return error

File: edge/pkg/edgehub/process.go (hunk 4 above)
Comment (fisherxu) [reply edge/pkg/edgehub/process.go]: If err != nil,  dispatch msg should be failed.

File: edge/pkg/edgehub/common/msghandler/handler.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		if handle.Filter(&message) {
+			err := handle.Process(&message, client)
+			if err != nil {
+				return false, fmt.Errorf("failed to handle message, error: %+v", err)
+			}
+			return true, nil
+		}
+	}
+	return true, nil
+}
```
Comment (fisherxu) [inline edge/pkg/edgehub/common/msghandler/handler.go:45]:
```suggestion
func ProcessHandler(message model.Message, client clients.Adapter) error {
	lock.RLock()
	defer lock.RUnlock()
	for _, handle := range Handlers {
		if handle.Filter(&message) {
			err := handle.Process(&message, client)
			if err != nil {
				return fmt.Errorf("failed to handle message, message group: %s, error: %+v", message.GetGroup(), err)
			}
			return nil
		}
	}
	return fmt.Errorf("failed to handle message, no handler found for the message, message group: %s", message.GetGroup())
}
```

File: edge/pkg/edgehub/process.go [hunk 9]
```diff
@@ -53,26 +68,34 @@ func (eh *EdgeHub) dispatch(message model.Message) error {
...
+func (eh *EdgeHub) dispatch(message model.Message) error {
+	// handler for msg.
+	isProcessed, err := msghandler.ProcessHandler(message, eh.chClient)
+	if isProcessed {
+		return nil
+	}
+
+	klog.Errorf("msg not processed due to not-supported, group: %v", message.GetGroup())
+	return fmt.Errorf("msg not processed due to not-supported, group: %v", message.GetGroup())
+}
```
Comment (fisherxu) [inline edge/pkg/edgehub/process.go:97]:
```suggestion
func (eh *EdgeHub) dispatch(message model.Message) error {
	// handler for msg.
	err := msghandler.ProcessHandler(message, eh.chClient)
	if err!=nil {
		return err
	}
        
        return nil
}
```

**Comment by fisherxu** [comment]
Please squash the commits, thanks.

**Comment by gy95** [comment]
> Please squash the commits, thanks.

done， PTAL

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**